### PR TITLE
Review comments for Data View selection support

### PR DESCRIPTION
**Pull request by kitchoi**
[Targeting #574]

This PR addresses and/or accompanies my own review comments on #574
As I was reviewing the PR, I find it necessary to test out my comments and suggestions. As I do so, I find more things in the PR that I otherwise would not have noticed just by reading it. I also got to weed out some of my faulty comments so no ones get to be bothered by them :)

This does not need to be merged, but I will let @corranwebster decide what to do. I did not try to break this down into smaller PRs because it is otherwise too hard.

Main changes:
- Validating row and column indexes even when the list of selections is mutated
- Resetting selection when the selection mode changes regardless of toolkit backend
- Handle when changing selection leads to changes to other things that will then change the selection
- Validating index for when selection type is "item"

New tests are added for all of these changes.

There is a new test case that depends on a fake data model.  This is needed to test the many code branches for validating row and column. Otherwise some of the edge cases cannot be covered simply by using the ArrayDataModel (which is a regular data structure).

In order to test row and column index in a toolkit agnostic fashion (the code itself is toolkit agnostic), I also had to expand the null toolkit's `Widget`. That in itself could be a separate PR.

### CODE REVIEW

File: pyface/data_view/tests/test_data_view_widget.py [hunk 1]
```diff
@@ -185,6 +502,52 @@ def test_selection_mode_extended(self):
...
+        # switch from "extended" to "single"
+        self.widget.selection = [((0,), ()), ((1,), ())]
+        self.gui.process_events()
+        self.assertEqual(self.widget._get_control_selection_mode(), "single")
+
+        if is_wx:
+            # Not sure why the selection is not updated.
+            self.assertEqual(
+                self.widget._get_control_selection(), [((0,), ()), ((1,), ())]
+            )
```
Comment (kitchoi) [inline pyface/data_view/tests/test_data_view_widget.py:533]: The expected value is an empty list. I tried to understand why this did not work for wx, but I couldn't.

File: pyface/data_view/i_data_view_widget.py [hunk 2]
```diff
@@ -116,7 +222,7 @@ def _set_control_selection_mode(self, selection_mode):
 
     def _selection_updated(self, event):
         """ Observer for selection trait. """
-        if self.control is not None and not self._selection_updating_flag:
```
Comment (kitchoi) [inline pyface/data_view/i_data_view_widget.py:119]: I removed this to test if it is necessary at all. I tried to test but I could not see why this is needed.

File: pyface/data_view/i_data_view_widget.py [hunk 3]
```diff
@@ -24,6 +27,108 @@
...
+            raise TraitError("Invalid row index {!r}".format(row))
+
+        if object.selection_type == 'item':
+            can_have_children = object.data_model.can_have_children(row)
+            have_rows = object.data_model.get_row_count(row) > 0
+            if can_have_children and have_rows:
+                raise TraitError(
+                    "Row values must have no children when selection_type "
+                    "is 'item', got {!r}".format(row)
+                )
```
Comment (kitchoi) [inline pyface/data_view/i_data_view_widget.py:52]: I added the validation for 'item' following the rest of the validation logic. But I have not added the code in the toolkit specific widget to actually avoid setting an invalid selection. e.g. if the user tries to click on row and column, the toolkit specific widget at the moment does not filter the event. Then we will get a trait error, and in Qt case, a hard crash. That filtering will need to be done if we include this validation.

File: pyface/data_view/tests/test_data_view_widget.py (hunk 1 above)
Comment (kitchoi) [reply pyface/data_view/tests/test_data_view_widget.py:533]: Weirdly, this assertion fails on Travis CI OSX build, but passes on Travis CI Linux build.
The assertion passes (i.e. expected failure) on my local machine too... It'd be good to get an independent confirmation. :puzzled:

Comment (kitchoi) [reply pyface/data_view/tests/test_data_view_widget.py:533]: On Travis CI linux build, it printed this error message and the test continued to "pass"(?):
```
test_selection_mode_change_when_selection_change (pyface.data_view.tests.test_data_view_widget.TestWidget) ... Traceback (most recent call last):
  File "/Users/travis/.edm/envs/pyface-test-3.6-wx/lib/python3.6/site-packages/pyface-7.1.0.dev110-py3.6.egg/pyface/data_view/i_data_view_widget.py", line 247, in _update_selection
    self.selection = new
  File "/Users/travis/.edm/envs/pyface-test-3.6-wx/lib/python3.6/site-packages/pyface-7.1.0.dev110-py3.6.egg/pyface/data_view/i_data_view_widget.py", line 127, in validate
    "is 'single', got {!r}".format(value)
traits.trait_errors.TraitError: Selection must have at most one element when selection_mode is 'single', got [((0,), ()), ((1,), ())]
Traceback (most recent call last):
  File "/Users/travis/.edm/envs/pyface-test-3.6-wx/lib/python3.6/site-packages/pyface-7.1.0.dev110-py3.6.egg/pyface/data_view/i_data_view_widget.py", line 247, in _update_selection
    self.selection = new
  File "/Users/travis/.edm/envs/pyface-test-3.6-wx/lib/python3.6/site-packages/pyface-7.1.0.dev110-py3.6.egg/pyface/data_view/i_data_view_widget.py", line 127, in validate
    "is 'single', got {!r}".format(value)
traits.trait_errors.TraitError: Selection must have at most one element when selection_mode is 'single', got [((0,), ()), ((1,), ())]
ok
```

File: pyface/data_view/i_data_view_widget.py (hunk 3 above)
Comment (kitchoi) [reply pyface/data_view/i_data_view_widget.py:52]: Or I might have misunderstood what "item" means. The above code tries to exclude row and column headers, but maybe "item" is meant to include them? Headers do tend to represent different data though. 🤫

File: pyface/data_view/tests/test_data_view_widget.py [hunk 4]
```diff
@@ -185,6 +502,52 @@ def test_selection_mode_extended(self):
...
+        self.addCleanup(
+            self.widget.observe,
+            change_selection_mode, "selection.items", dispatch="ui",
+            remove=True,
+        )
+
+        # switch from "extended" to "single"
+        self.widget.selection = [((0,), ()), ((1,), ())]
+        self.gui.process_events()
+        self.assertEqual(self.widget._get_control_selection_mode(), "single")
```
Comment (corranwebster) [inline pyface/data_view/tests/test_data_view_widget.py:527]: Just to clarify, this is not how selection mode works! Selection mode tells you what sorts of things can be selected, and does not change based on what the actual selection _is_.

When selection mode is `single` then the selection is always a list with one item.

When selection mode is `extened` then the selection is a list with any number of items, _including_ possibly just one item.

Comment (kitchoi) [reply pyface/data_view/tests/test_data_view_widget.py:527]: I am confused. From reading the code, the widget resets the selection to an empty list whenever selection mode is changed.
The line you are commenting on is asserting the selection *mode*. When the selection mode is changed on the `self.widget`, the selection mode on the GUI control should be synchronized to the new value, right?

Comment (kitchoi) [reply pyface/data_view/tests/test_data_view_widget.py:527]: To clarify, this test is testing the logic in this method:
https://github.com/enthought/pyface/blob/9daeb87e832e85d4ce0fbd29a5dc1a700bdec3bb/pyface/ui/qt4/data_view/data_view_widget.py#L91-L95

**Comment by corranwebster** [comment]
Closing, as the original PR is merged.

**Comment by corranwebster** [comment]
Ooops - underlying PR hasn't been merged, but still closing as things have been fixed independently in that PR.

**Comment by kitchoi** [comment]
Yep. Thank you!
